modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/affinity/PartitionExtractor.java

```diff
@@ -366,4 +379,171 @@ else if (rightParam != null)
+        if (!((leftOpType == GridSqlOperationType.BIGGER || leftOpType == GridSqlOperationType.BIGGER_EQUAL)
+            && (rightOpType == GridSqlOperationType.SMALLER || rightOpType == GridSqlOperationType.SMALLER_EQUAL)))
+            return null;
+
+        // Try parse left AST right value (value to the right of '>' or '>=').
+        GridSqlAst leftVal = leftOp.child(1);
+
+        GridSqlConst leftConst;
+        GridSqlParameter leftParam;
+
+        if (leftVal instanceof GridSqlConst) {
+            leftConst = (GridSqlConst)leftVal;
+            leftParam = null;
+        }
+        else if (leftVal instanceof GridSqlParameter) {
+            leftConst = null;
+            leftParam = (GridSqlParameter)leftVal;
+        }
+        else
+            return null;
+
+        // Try parse right AST right value (value to the right of '<' or '<=').
+        GridSqlAst rightVal = rightOp.child(1);
+
+        GridSqlConst rightConst;
+        GridSqlParameter rightParam;
+
+        if (rightVal instanceof GridSqlConst) {
+            rightConst = (GridSqlConst)rightVal;
+            rightParam = null;
+        }
+        else if (rightVal instanceof GridSqlParameter) {
+            rightConst = null;
+            rightParam = (GridSqlParameter)rightVal;
+        }
+        else
+            return null;
+
+        if (leftParam != null || rightParam != null)
+            return null;
+
+        // Check const dataTypes
+        int leftConstType = leftConst.value().getType();
+
+        if (!(leftConstType == Value.BYTE || leftConstType == Value.SHORT || leftConstType == Value.INT ||
+            leftConstType == Value.LONG))
+            return null;
+
+        int rightConstType = rightConst.value().getType();
+
+        if (!(rightConstType == Value.BYTE || rightConstType == Value.SHORT || rightConstType == Value.INT ||
+            rightConstType == Value.LONG))
+            return null;
+
+
+        long leftLongVal = leftConst.value().getLong();
+        long rightLongVal = rightConst.value().getLong();
+
+        // Increment left long value if '>' is used.
+        if (leftOpType == GridSqlOperationType.BIGGER)
+            leftLongVal++;
+
+        // Decrement right long value if '<' is used.
+        if (rightOpType == GridSqlOperationType.SMALLER)
+            rightLongVal--;
+
+        Set<PartitionSingleNode> parts = new HashSet<>();
+
+        for (long i = leftLongVal; i <= rightLongVal; i++) {
+            // Do extract.
+            PartitionSingleNode part = extractSingle(leftCol.column(), new GridSqlConst(ValueLong.get(i)),
+                null);
+
+            if (part == null)
+                return null;
+
+            parts.add(part);
+
+            int maxPartitionsCnt = Integer.getInteger(
+                IgniteSystemProperties.IGNITE_PARTITIONS_PRUNNING_MAX_PARTIONS_BETWEEN, MAX_PARTITIONS_COUNT_BETWEEN);
+
+            if (parts.size() > maxPartitionsCnt)
```

Comment:
I wonder why do we bound by accumulated partitions count but not by range between `leftLongVal` and `rightLongVal`?

Reply:
Cause both leftLongVal and rightLongVal are keys and not partitions and different keys might match to same partitions. For example (between 1023 and 1025) might match to two partitions.